tools/nni_trial_tool/trial_runner.py

```diff
@@ -0,0 +1,264 @@
+    args.enable_gpu_collect = settings["enableGpuCollector"]
+    args.command_channel = settings["commandChannel"]
+
+    if args.trial_command is None:
+        args.trial_command = settings["command"]
+    if args.nnimanager_ip is None:
+        args.nnimanager_ip = settings["nniManagerIP"]
+    if args.nnimanager_port is None:
+        args.nnimanager_port = settings["nniManagerPort"]
+    if args.nni_manager_version is None:
+        args.nni_manager_version = settings["nniManagerVersion"]
+    if args.log_collection is None:
+        args.log_collection = settings["logCollection"]
+    if args.node_count is None:
+        # default has only one node.
+        args.node_count = 1
+
+    os.environ['NNI_OUTPUT_DIR'] = os.curdir + "/nnioutput"
+    os.environ['NNI_PLATFORM'] = args.platform
+    os.environ['NNI_SYS_DIR'] = os.curdir
+    os.environ['NNI_EXP_ID'] = args.exp_id
+    os.environ['MULTI_PHASE'] = "true"
+    os.environ['NNI_TRIAL_JOB_ID'] = "runner"
+
+    from .log_utils import LogType, RemoteLogger, StdOutputType, nni_log
+    from .rest_utils import rest_get, rest_post
+    from .url_utils import gen_parameter_meta_url, gen_send_version_url
+    from .trial import Trial
+    from .file_channel import FileChannel
+    from .web_channel import WebChannel
+    from .base_channel import CommandType
+
+    is_multi_node = args.node_count > 1
+
+    if (is_multi_node):
+        # for multiple nodes, create a file to get a unique id.
+        while True:
+            node_id = random.randint(0, 10000)
+            unique_check_file_name = "node_%s" % (node_id)
+            if not os.path.exists(unique_check_file_name):
+                break
+        with open(unique_check_file_name, "w") as unique_check_file:
+            unique_check_file.write("%s" % (int(datetime.now().timestamp() * 1000)))
+        args.node_id = node_id
+    else:
+        # node id is unique in the runner
+        args.node_id = None
+
+    trial_runner_syslogger = RemoteLogger(args.nnimanager_ip, args.nnimanager_port, 'runner',
+                                          StdOutputType.Stdout, args.log_collection, args.runner_name)
+    sys.stdout = sys.stderr = trial_runner_syslogger
+    nni_log(LogType.Info, "{}: merged args is {}".format(args.node_id, args))
+
+    if args.trial_command is None:
+        nni_log(LogType.Error, "{}: no command is found.".format(args.node_id))
+        os._exit(1)
+    check_version(args)
+    try:
+        main_loop(args)
+    except SystemExit as se:
+        nni_log(LogType.Info, '{}: NNI trial runner exit with code {}'.format(args.node_id, se.code))
+        os._exit(se.code)
+    finally:
+        if trial_runner_syslogger is not None:
+            if trial_runner_syslogger.pipeReader is not None:
+                trial_runner_syslogger.pipeReader.set_process_exit()
+            trial_runner_syslogger.close()
+
+    # the process doesn't exit even main loop exit. So exit it explictly.
```

Comment:
why the process does not exit?

Reply:
in some error cases, subthreads not exit. So it needs to force exit to prevent this kind of situation.